integration-tests/deployment/keystone/deploy_test.go

```diff
@@ -0,0 +1,246 @@
+		// assign nops to nodes
+		for _, env := range e.DonToEnv {
+			for i, nodeID := range env.NodeIDs {
+				idx := i % len(testNops)
+				nop := testNops[idx]
+				nodeToNop[nodeID] = nop
+			}
+		}
+
+		var donsToDeploy = map[string][]kcr.CapabilitiesRegistryCapability{
+			keystone.WFDonName:     []kcr.CapabilitiesRegistryCapability{keystone.OCR3Cap},
+			keystone.TargetDonName: []kcr.CapabilitiesRegistryCapability{keystone.WriteChainCap},
+		}
+
+		ctx := context.Background()
+		// Deploy all the Keystone contracts.
+		homeChain := e.Get(keystone.WFDonName).AllChainSelectors()[0]
+		deployReq := keystone.DeployRequest{
+			RegistryChain:     homeChain,
+			Menv:              e,
+			DonToCapabilities: donsToDeploy,
+			NodeIDToNop:       nodeToNop,
+		}
+
+		deployResp, err := keystone.Deploy(ctx, lggr, deployReq)
+		require.NoError(t, err)
+		ad := deployResp.Changeset.AddressBook
+		addrs, err := ad.Addresses()
+		require.NoError(t, err)
+		lggr.Infow("Deployed Keystone contracts", "address book", addrs)
+
+		// all contracts on home chain
+		homeChainAddrs, err := ad.AddressesForChain(homeChain)
+		require.NoError(t, err)
+		require.Len(t, homeChainAddrs, 3)
+		// only forwarder on non-home chain
+		for _, chain := range e.Get(keystone.TargetDonName).AllChainSelectors() {
+			chainAddrs, err := ad.AddressesForChain(chain)
+			require.NoError(t, err)
+			if chain != homeChain {
+				require.Len(t, chainAddrs, 1)
+			} else {
+				require.Len(t, chainAddrs, 3)
+			}
+			containsForwarder := false
+			for _, tv := range chainAddrs {
+				if tv.Type == keystone.KeystoneForwarder {
+					containsForwarder = true
+					break
+				}
+			}
+			require.True(t, containsForwarder, "no forwarder found in %v on chain %d for target don", chainAddrs, chain)
+		}
+
+		req := &keystone.GetContractSetsRequest{
+			Chains:      e.Chains(),
+			AddressBook: ad,
+		}
+
+		contractSetsResp, err := keystone.GetContractSets(lggr, req)
+		require.NoError(t, err)
+		require.Len(t, contractSetsResp.ContractSets, 4)
+		// check the registry
+		regChainContracts, ok := contractSetsResp.ContractSets[homeChain]
+		require.True(t, ok)
+		gotRegistry := regChainContracts.CapabilitiesRegistry
+		require.NotNil(t, gotRegistry)
+		// contract reads
+		gotDons, err := gotRegistry.GetDONs(nil)
+		if err != nil {
+			err = keystone.DecodeErr(kcr.CapabilitiesRegistryABI, err)
+			require.Fail(t, "failed to get Dons", err)
+		}
+		require.NoError(t, err)
+		assert.Len(t, gotDons, len(e.DonToEnv))
+		for don, id := range deployResp.DonToId {
+			// id starts at 1 in the contract
+			gdon := gotDons[id-1]
+			cfg, ok := multDonCfg.Configs[don]
+			require.True(t, ok, "no config for don %s", don)
+			assert.Equal(t, cfg.Nodes/3, int(gdon.F))
+			assert.Len(t, gdon.NodeP2PIds, cfg.Nodes)
+			assert.Equal(t, don == keystone.WFDonName, gdon.AcceptsWorkflows, "don %s, %d has wrong AcceptsWorkflows", don, id)
+		}
+	})
+
+	t.Run("memory chains clo offchain", func(t *testing.T) {
```

Comment:
this test mimic what will happen for a real set of testnets.

the difference is hidden in `NewDonEnvWithMemoryChains`, which allocates a simulated backend for each chain that the node supports.

on testnet, we will inject real chains.